ratelimit/record_handlers/helpers.py

```diff
@@ -0,0 +1,123 @@
+from datetime import timedelta
+from hashlib import md5
+from typing import Callable, Optional
+
+import ipware.ip
+from django.core.cache import BaseCache, caches
+from django.utils.module_loading import import_string
+from ratelimit.conf import settings
+
+
+def get_client_ip_address(request) -> str:
+    """
+    Get client IP address from request.
+
+    The django-ipware package is used for address resolution.
+    """
+
+    client_ip_address, _ = ipware.ip.get_client_ip(request)
+    return client_ip_address
+
+
+def get_client_user_agent(request) -> str:
+    return request.META.get("HTTP_USER_AGENT", "<unknown>")[:255]
+
+
+def get_client_path_info(request) -> str:
+    return request.META.get("PATH_INFO", "<unknown>")[:255]
+
+
+def get_client_username(request) -> str:
+    if request.user.is_authenticated:
+        return request.user.username
+    return "Anonymous"
```

Comment:
Not everyone has a "username" field. You might want to look into `django.contrib.auth.models` where they define the Username field inside a list (specifically AbstractUser or AbstractBaseUser). I also think this should instead be based on user ID/pk instead of username to be more efficient (string lookup vs integer lookup).

If you're going to stick with username, then try this (but I'd still recommend using the integer pk AutoField):
```suggestion
from django.contrib.auth.models import AnonymousUser

def get_client_username(request) -> str:
    if request.user.is_authenticated:
        return getattr(request.user, request.user.USERNAME_FIELD)
    return AnonymousUser
```